text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1020 @@
+to users, since Ember has not had any modifiers besides `{{action}}`, and the
+element modifier manager RFC was just merged recently. Key things to focus on
+here will be the fact the modifiers run whenever a node is inserted or removed,
+so they may run multiple times during a components life (if they exist in an
+`{{if}}` or `{{each}}` block, for instance), and that they can receive pure
+function helpers if they do not require any state beyond the state of the
+element itself, which can simplify component code.
+
+### Converting From Classic Components
+
+For users who are familiar with the classic component API, some changes will be
+fairly simple, and others will be more nuanced. The individual focus points
+should be those outlined in the motivation of this RFC, and the motivation for
+the changes should be included in a summary that helps users understand the
+intent of new APIs so they can use them effectively.
+
+**Outer HTML semantics** should be fairly simple to communicate. These should
+also be the easiest to codemod, since it is mostly a 1-to-1 conversion from the
+classic component definition to the template.
+
+**Namespaced arguments** will be a sticking point for some. The main thing to
+focus on here is aliasing to provide defaults, and the benefits of immutability.
+Patterns like the `@arg` decorator in
+[sparkles-decorators](https://github.com/gossi/sparkles-decorators) make this
+fairly ergonomic, and for simpler components defaulting using a class field
+initializer can also work.
+
+**Lifecycle hooks** may be fairly confusing to existing users, who are used to
+having a variety of hooks to choose from when coordinating lifecycle events. The
+bullet points here are:
+
+* Tracked properties/computed properties are the primary place to react to
+  argument changes for any values that can be computed directly via getters.
+  Ideally, most logic for derived state is conventionally in tracked or computed
+  properties.
+* `didUpdateArgs` is the place to react to argument changes that don't fit into
+  getters. This includes asynchronous calls, conditional changes (e.g. if an arg
+  changed, do X, else do nothing), etc.
+* `willDestroy` is the correct place for all teardown code, like in classic
+  components.
+* `didReceiveAttrs`, `willRender`, and `didRender` code should be extracted into
+  functions which are then passed to `{{did-render}}`
+* `willDestroyElement` code should be extracted into functions which are then
+  passed to `{{will-destroy}}`
+
+**Element modifiers** will be new to almost everyone, and this will have to be
+a primary focus in learning. The bullet points for these are:
+
+* They are conceptually like the render lifecycle hooks on classic components,
+  but they run on elements instead of components.
+* `{{did-render}}` is similar to the `didRender` lifecycle hooks in classic
+  components. It will run when the element is inserted, and then it will run
+  again when any of its _args_ change.
```

Comment:
I have some concerns about the relative complexity of patterns like the twitter bootstrap 3 tabs. It may grow significantly relative to the classic `didInsertElement` way of doing things.

Developers create a HTML structure like

```html
<!-- Nav tabs -->
  <ul class="nav nav-tabs" role="tablist">
    <li role="presentation" class="active"><a href="#home" aria-controls="home" role="tab" data-toggle="tab">Home</a></li>
    <li role="presentation"><a href="#profile" aria-controls="profile" role="tab" data-toggle="tab">Profile</a></li>
    <li role="presentation"><a href="#messages" aria-controls="messages" role="tab" data-toggle="tab">Messages</a></li>
    <li role="presentation"><a href="#settings" aria-controls="settings" role="tab" data-toggle="tab">Settings</a></li>
  </ul>

  <!-- Tab panes -->
  <div class="tab-content">
    <div role="tabpanel" class="tab-pane active" id="home">...</div>
    <div role="tabpanel" class="tab-pane" id="profile">...</div>
    <div role="tabpanel" class="tab-pane" id="messages">...</div>
    <div role="tabpanel" class="tab-pane" id="settings">...</div>
  </div>
```
and then run (w/ optional jQuery enabled)
```js
$('#myTabs a').click(function (e) {
  e.preventDefault()
  $(this).tab('show')
})
```

We have two sibling elements that sort of relate to eachother, and are "activated" by some foreign JS plugin. It doesn't feel entirely correct to put a modifier-based hook on either `<ul class="nav nav-tabs" role="tablist">` or `<div class="tab-content">` since it's the pair of them that's important, and neither is the parent of the other.

Is our answer here to add a wrapping element for the sole purpose of having a common parent of the two siblings which must be set-up/destroyed together?

Reply:
I think in this case it would make sense to place the modifier on just the tabs, since they are controlling the content, and the modifier would be setting up the control. A wrapping div would not be necessary, though if users felt it made more sense they could use it.